File: solutions/0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string str) {
        // Input sanity checks.
        if(str.length() == 0) {
            return true;
        }
        if((str.length() % 2) != 0) {
            return false;
        }
        // Sanity check, if the input string contains any other char than the below six.
        // Return false.
        std::stack<char> cmp;
        char open_param = '('; char close_param = ')';
        char open_brack = '{'; char close_brack = '}';
        char open_sq = '['; char close_sq = ']';
        for (size_t i=0; i < str.length(); i++)
        {
            if (str.at(i) == open_param ||
                str.at(i) == open_brack ||
                str.at(i) == open_sq)
            {
                cmp.push(str.at(i));
                continue;
            }
            if (cmp.empty()) {
                return false;
            }
            char top = cmp.top();
            bool check_param_fail =  (top == open_param && str.at(i) != close_param) ||
                                (top == open_sq && str.at(i) != close_sq) ||
                                (top == open_brack && str.at(i) != close_brack);
            if (check_param_fail) {
                return false;
            }
            cmp.pop();
        }
        if (!cmp.empty())
            return false;
        return true;        
    }
};

File: solutions/0020-valid-parentheses/solution.go
import "github.com/emirpasic/gods/stacks/arraystack"

const (
    open_param = '('
    close_param = ')'
    open_sq = '['
    close_sq = ']'
    open_fl = '{'
    close_fl = '}'
)

func isValid(s string) bool {
    if (len(s) == 0) {
        return true
    }
    if (len(s) % 2) != 0 {
        return false
    }
    stck := arraystack.New()
    for _, c := range s {
        is_open :=  (c == open_param ||
                     c == open_sq ||
                     c == open_fl)
        if is_open {
            stck.Push(c)
            continue
        }
        if stck.Empty() {
            return false
        }
        top, _ := stck.Pop()
        is_mismatched := (top == open_param && c != close_param ||
                          top == open_sq && c != close_sq ||
                          top == open_fl && c != close_fl)
        if is_mismatched {
            return false
        }
    }
    return stck.Empty()
}
